File: source/usvariablecommands.ts
// unit usvariablecommands

import { int, compareTextIgnoreCase, Color } from "./common"
import { KfCommand } from "./KfCommand"
import { TSVariable, TSVariableState } from "./TSVariable"
import { TSChangedVariableWrapper } from "./TSChangedVariableWrapper"
import { TSRule } from "./TSRule"

// TODO: FIX THIS
const usconsoleform: any = {};
const usdomain: any = {};
const usruleeditorform: any = {};

/*
from conversion_common import *
import usruleeditorform
import usdomain
import usconsoleform
import usworld
import ucommand
import delphi_compatability
*/

export class TSToggleVariableCommand extends KfCommand {
    variable: TSVariable = new TSVariable()
    oldState: TSVariableState
    newState: TSVariableState
    
    // ----------------------------- TSToggleVariableCommand --------------------- 
    constructor(variable: TSVariable) {
        super()
        this.variable = variable
        this.oldState = variable.getState()
        if (this.oldState === TSVariableState.kPresent) {
            this.newState = TSVariableState.kAbsent
        } else {
            this.newState = TSVariableState.kPresent
        }
    }
    
    setVariableStateWithUpdate(state: TSVariableState): void {
        this.variable.setState(state)
        if (usconsoleform.ConsoleForm.ShowOnlyTrueVariablesButton.Down) {
            usconsoleform.ConsoleForm.updateVariables()
            usconsoleform.ConsoleForm.VariablesListBox.ItemIndex = usconsoleform.ConsoleForm.VariablesListBox.Items.IndexOfObject(this.variable)
            usconsoleform.ConsoleForm.VariablesListBox.Invalidate()
        } else {
            usconsoleform.ConsoleForm.VariablesListBox.Invalidate()
        }
        usdomain.domain.world.updateAvailable()
        usconsoleform.ConsoleForm.speechSystem.listenForAvailableCommands()
    }
    
    doCommand(): void {
        this.setVariableStateWithUpdate(this.newState)
        KfCommand.prototype.doCommand.call(this)
    }
    
    undoCommand(): void {
        this.setVariableStateWithUpdate(this.oldState)
        KfCommand.prototype.undoCommand.call(this)
    }
    
    description(): string {
        let result = ""
        if (this.newState === TSVariableState.kPresent) {
            result = "toggle \"" + this.variable.phrase + "\" to true"
        } else {
            result = "toggle \"" + this.variable.phrase + "\" to false"
        }
        return result
    }
    
}

// need to have abstract base so TSDoCommandPhrase can defer updating till after changes
export class TSAbstractMoveFocusCommand extends KfCommand {
    oldFocus: TSVariable
    oldFocusOldState: TSVariableState
    newFocus: TSVariable
    newFocusOldState: TSVariableState
    
    // ----------------------------- TSAbstractMoveFocusCommand --------------------- 
    constructor(newFocus: TSVariable) {
        super()
        // the old states are stored for undo in case author has been toggling them individually
        this.newFocus = newFocus
        this.newFocusOldState = newFocus.getState()
        if (usdomain.domain.world.focus !== null) {
            this.oldFocus = usdomain.domain.world.focus
            this.oldFocusOldState = this.oldFocus.getState()
        } else {
            this.oldFocus = newFocus
            this.oldFocusOldState = newFocus.getState()
        }
    }
    
    updateForChanges(): void {
        usdomain.domain.world.updateAvailable()
        usconsoleform.ConsoleForm.speechSystem.listenForAvailableCommands()
        usconsoleform.ConsoleForm.updateVariables()
        usconsoleform.ConsoleForm.VariablesListBox.Invalidate()
    }
    
    shiftsFocus(): boolean {
        let result = false
        result = (this.newFocus !== usdomain.domain.world.emptyEntry) && (this.newFocus !== this.oldFocus)
        return result
    }
    
}

export class TSMoveFocusCommand extends TSAbstractMoveFocusCommand {
    
    // ----------------------------- TSMoveFocusCommand --------------------- 
    doCommand(): void {
        this.oldFocus.setState(TSVariableState.kAbsent)
        usdomain.domain.world.focus = this.newFocus
        this.newFocus.setState(TSVariableState.kPresent)
        this.updateForChanges()
        TSAbstractMoveFocusCommand.prototype.doCommand.call(this)
    }
    
    undoCommand(): void {
        this.newFocus.setState(this.newFocusOldState)
        usdomain.domain.world.focus = this.oldFocus
        this.oldFocus.setState(this.oldFocusOldState)
        this.updateForChanges()
        TSAbstractMoveFocusCommand.prototype.undoCommand.call(this)
    }
    
    description(): string {
        let result = ""
        result = "move focus to " + this.newFocus.phrase
        return result
    }
    
}

export class TSDoCommandPhrase extends TSAbstractMoveFocusCommand {
    commandPhrase: string
    changedVariables: TSChangedVariableWrapper[]
    oldLastSaidTextWithMacros: string
    newLastSaidTextWithMacros: string
    oldFirstCommandDoneForLastCommandPhrase: int
    newFirstCommandDoneForLastCommandPhrase: int
    
    // ----------------------------- TSDoCommandPhrase -------------------------------
    constructor(commandPhrase: string) {
        // determine what would need to change - including new focus and all variables
        // Reorganized with temporary variables to work around TypeScript limitation
        // of no access to this before super constructor called.
        const changedVariables: TSChangedVariableWrapper[] = []
        let newFocus = usdomain.domain.world.emptyEntry
        const oldLastSaidTextWithMacros = usconsoleform.ConsoleForm.speechSystem.lastSaidTextWithMacros
        let newLastSaidTextWithMacros = ""
        const oldFirstCommandDoneForLastCommandPhrase = usdomain.domain.world.firstCommandDoneForLastCommandPhrase
        let newFirstCommandDoneForLastCommandPhrase = -1
        for (let i = 0; i < usdomain.domain.world.rules.length; i++) {
            const rule: TSRule = usdomain.domain.world.rules[i]
            if (rule.available && compareTextIgnoreCase(rule.command.phrase, commandPhrase)) {
                usruleeditorform.RuleEditorForm.lastCommand = rule
                if (newFirstCommandDoneForLastCommandPhrase === -1) {
                    newFirstCommandDoneForLastCommandPhrase = i
                }
                const ruleResult = rule.recordReplyMoveChanges(changedVariables, newLastSaidTextWithMacros)
                newLastSaidTextWithMacros = ruleResult.totalReply
                if (ruleResult.contextToFocusTo) {
                    newFocus = ruleResult.contextToFocusTo
                }
            }
        }

        if (commandPhrase.startsWith("$")) {
            // elimitate leading $
            commandPhrase = commandPhrase.substring(1)
        }

        super(newFocus)

        this.newFocus = newFocus
        this.oldLastSaidTextWithMacros = oldLastSaidTextWithMacros
        this.newLastSaidTextWithMacros = newLastSaidTextWithMacros
        this.oldFirstCommandDoneForLastCommandPhrase  = oldFirstCommandDoneForLastCommandPhrase 
        this.newFirstCommandDoneForLastCommandPhrase = newFirstCommandDoneForLastCommandPhrase
        this.commandPhrase = commandPhrase
        this.changedVariables = changedVariables
    }
    
    doCommand(): void {
        let i: int
        
        usconsoleform.ConsoleForm.addLineToTranscript("> " + this.commandPhrase, Color.clRed)
        usconsoleform.ConsoleForm.addLineToTranscript(usconsoleform.ConsoleForm.speechSystem.stripMacros(this.newLastSaidTextWithMacros), Color.clBlue)
        usconsoleform.ConsoleForm.scrollTranscriptEndIntoView()
        usconsoleform.ConsoleForm.speechSystem.sayTextWithMacros(this.newLastSaidTextWithMacros)
        //common with undo
        usconsoleform.ConsoleForm.speechSystem.lastSaidTextWithMacros = this.newLastSaidTextWithMacros
        usdomain.domain.world.firstCommandDoneForLastCommandPhrase = this.newFirstCommandDoneForLastCommandPhrase
        if (this.newFocus !== usdomain.domain.world.emptyEntry) {
            this.oldFocus.setState(TSVariableState.kAbsent)
            usdomain.domain.world.focus = this.newFocus
            this.newFocus.setState(TSVariableState.kPresent)
        }
        for (i = 0; i < this.changedVariables.length; i++) {
            this.changedVariables[i].doChange()
        }
        this.updateForChanges()
        usconsoleform.ConsoleForm.speechSystem.checkForSayOptionsMacro()
        super.doCommand()
    }
    
    undoCommand(): void {
        let i: int
        let undoPhrase: string
        
        usconsoleform.ConsoleForm.addLineToTranscript("> undo", Color.clRed)
        //  undoPhrase := 'It is as if "' + commandPhrase + '" had never been said.';
        undoPhrase = "(You decide not to say \"" + this.commandPhrase + "\")"
        usconsoleform.ConsoleForm.addLineToTranscript(undoPhrase, Color.clBlue)
        usconsoleform.ConsoleForm.scrollTranscriptEndIntoView()
        usconsoleform.ConsoleForm.speechSystem.speakText(undoPhrase)
        usconsoleform.ConsoleForm.speechSystem.lastSaidTextWithMacros = this.oldLastSaidTextWithMacros
        usdomain.domain.world.firstCommandDoneForLastCommandPhrase = this.oldFirstCommandDoneForLastCommandPhrase
        for (let i = this.changedVariables.length - 1; i >= 0; i--) {
            this.changedVariables[i].undoChange()
        }
        if (this.newFocus !== usdomain.domain.world.emptyEntry) {
            this.newFocus.setState(this.newFocusOldState)
            usdomain.domain.world.focus = this.oldFocus
            this.oldFocus.setState(this.oldFocusOldState)
        }
        this.updateForChanges()
        super.undoCommand()
    }
    
    redoCommand(): void { 
        usconsoleform.ConsoleForm.addLineToTranscript("> redo", Color.clRed)
        const redoPhrase = "(You decide to say \"" + this.commandPhrase + "\" anyway)"
        usconsoleform.ConsoleForm.addLineToTranscript(redoPhrase, Color.clBlue)
        usconsoleform.ConsoleForm.scrollTranscriptEndIntoView()
        usconsoleform.ConsoleForm.speechSystem.speakText(redoPhrase)
        //common with do
        usconsoleform.ConsoleForm.speechSystem.lastSaidTextWithMacros = this.newLastSaidTextWithMacros
        usdomain.domain.world.firstCommandDoneForLastCommandPhrase = this.newFirstCommandDoneForLastCommandPhrase
        if (this.newFocus !== usdomain.domain.world.emptyEntry) {
            this.oldFocus.setState(TSVariableState.kAbsent)
            usdomain.domain.world.focus = this.newFocus
            this.newFocus.setState(TSVariableState.kPresent)
        }
        for (let i = 0; i < this.changedVariables.length; i++) {
            this.changedVariables[i].doChange()
        }
        this.updateForChanges()
        super.doCommand()
    }
    
    description(): string {
        let result = ""
        result = "command: " + this.commandPhrase
        return result
    }
    
}

File: converted_source/used/usvariablecommands.pas
unit usvariablecommands;

interface

uses Windows, SysUtils, Graphics, Classes, Controls, StdCtrls, UCommand, USWorld;

type

  TSToggleVariableCommand = class(KfCommand)
    variable: TSVariable;
    oldState: TSVariableState;
    newState: TSVariableState;
    constructor createWithVariable(variable: TSVariable);
		procedure setVariableStateWithUpdate(state: TSVariableState);
  	procedure doCommand; override;
  	procedure undoCommand; override;
  	function description: string; override;
    end;

  // need to have abstract base so TSDoCommandPhrase can defer updating till after changes
  TSAbstractMoveFocusCommand = class(KfCommand)
    oldFocus: TSVariable;
    oldFocusOldState: TSVariableState;
    newFocus: TSVariable;
    newFocusOldState: TsVariableState;
    constructor createWithNewFocus(newFocus: TSVariable);
		procedure updateForChanges;
    function shiftsFocus: boolean;
    end;

  TSMoveFocusCommand = class(TSAbstractMoveFocusCommand)
  	procedure doCommand; override;
  	procedure undoCommand; override;
  	function description: string; override;
    end;

  TSDoCommandPhrase = class(TSAbstractMoveFocusCommand)
    commandPhrase: string;
    changedVariables: TList;
    oldLastSaidTextWithMacros: string;
    newLastSaidTextWithMacros: string;
    oldFirstCommandDoneForLastCommandPhrase: integer;
    newFirstCommandDoneForLastCommandPhrase: integer;
    constructor createWithCommandPhrase(const commandPhrase: string);
		destructor destroy; override;
  	procedure doCommand; override;
  	procedure undoCommand; override;
 		procedure redoCommand; override;
  	function description: string; override;
    end;

implementation

uses USConsoleForm, USDomain, USRuleEditorForm;

{ ----------------------------- TSToggleVariableCommand --------------------- }

constructor TSToggleVariableCommand.createWithVariable(variable: TSVariable);
  begin
  self.create;
  self.variable := variable;
  self.oldState := variable.getState;
  if oldState = kPresent then
  	newState := kAbsent
  else
  	newState := kPresent;
  end;

procedure TSToggleVariableCommand.setVariableStateWithUpdate(state: TSVariableState);
  begin
  variable.setState(state);
  if ConsoleForm.ShowOnlyTrueVariablesButton.down then
    begin
    ConsoleForm.updateVariables;
    ConsoleForm.VariablesListBox.itemIndex := ConsoleForm.VariablesListBox.items.indexOfObject(variable);
  	ConsoleForm.VariablesListBox.invalidate;
    end
  else
    ConsoleForm.VariablesListBox.invalidate;
  domain.world.updateAvailable;
  ConsoleForm.speechSystem.listenForAvailableCommands;
	end;

procedure TSToggleVariableCommand.doCommand;
  begin
  self.setVariableStateWithUpdate(newState);
	inherited doCommand;
  end;

procedure TSToggleVariableCommand.undoCommand;
  begin
  self.setVariableStateWithUpdate(oldState);
  inherited undoCommand;
  end;

function TSToggleVariableCommand.description: string;
  begin
  if newState = kPresent then
  	result := 'toggle "' + variable.phrase + '" to true'
  else
    result := 'toggle "' + variable.phrase + '" to false';
  end;

{ ----------------------------- TSAbstractMoveFocusCommand --------------------- }

constructor TSAbstractMoveFocusCommand.createWithNewFocus(newFocus: TSVariable);
  begin
  self.create;
  // the old states are stored for undo in case author has been toggling them individually
  self.newFocus := newFocus;
  self.newFocusOldState := newFocus.getState;
  if domain.world.focus <> nil then
    begin
  	self.oldFocus := domain.world.focus;
  	self.oldFocusOldState := self.oldFocus.getState;
    end
  else
    begin
  	self.oldFocus := newFocus;
  	self.oldFocusOldState := newFocus.getState;
    end;
  end;

procedure TSAbstractMoveFocusCommand.updateForChanges;
  begin
  domain.world.updateAvailable;
  ConsoleForm.speechSystem.listenForAvailableCommands;
  ConsoleForm.updateVariables;
  ConsoleForm.VariablesListBox.invalidate;
	end;

function TSAbstractMoveFocusCommand.shiftsFocus: boolean;
  begin
  result := (newFocus <> domain.world.emptyEntry) and (newFocus <> oldFocus);
  end;

{ ----------------------------- TSMoveFocusCommand --------------------- }

procedure TSMoveFocusCommand.doCommand;
  begin
  oldFocus.setState(kAbsent);
  domain.world.focus := newFocus;
  newFocus.setState(kPresent);
  self.updateForChanges;
	inherited doCommand;
  end;

procedure TSMoveFocusCommand.undoCommand;
  begin
  newFocus.setState(newFocusOldState);
  domain.world.focus := oldFocus;
  oldFocus.setState(oldFocusOldState);
  self.updateForChanges;
  inherited undoCommand;
  end;

function TSMoveFocusCommand.description: string;
  begin
  result := 'move focus to ' + newFocus.phrase;
  end;

{ ----------------------------- TSDoCommandPhrase -------------------------------}

constructor TSDoCommandPhrase.createWithCommandPhrase(const commandPhrase: string);
  var
    i: integer;
    rule: TSRule;
  begin
  self.commandPhrase := commandPhrase;
  changedVariables := TList.create;
  // determine what would need to change - including new focus and all variables
  newFocus := domain.world.emptyEntry;
  oldLastSaidTextWithMacros := ConsoleForm.speechSystem.lastSaidTextWithMacros;
  newLastSaidTextWithMacros := '';
  oldFirstCommandDoneForLastCommandPhrase := domain.world.firstCommandDoneForLastCommandPhrase;

  newFirstCommandDoneForLastCommandPhrase := -1;
  if domain.world.rules.count > 0 then
    for i := 0 to domain.world.rules.count - 1 do
      begin
      rule := TSRule(domain.world.rules.items[i]);
      if rule.available and (AnsiCompareText(rule.command.phrase, commandPhrase) = 0) then
        begin
        RuleEditorForm.lastCommand := rule;
        if newFirstCommandDoneForLastCommandPhrase = -1 then
        	newFirstCommandDoneForLastCommandPhrase := i;
        rule.recordReplyMoveChanges(changedVariables, newLastSaidTextWithMacros, newFocus);
       	end;
      end;

  // elimitate leading $
  if (length(self.commandPhrase) > 1) and (self.commandPhrase[1] = '$') then
    self.commandPhrase := copy(self.commandPhrase, 2, length(self.commandPhrase));

  self.createWithNewFocus(newFocus);
  end;

destructor TSDoCommandPhrase.destroy;
  var
  	i: integer;
  begin
  for i := 0 to changedVariables.count - 1 do
    TSChangedVariableWrapper(changedVariables[i]).free;
  changedVariables.free;
  changedVariables := nil;
  inherited destroy;
  end;

procedure TSDoCommandPhrase.doCommand;
  var
  	i: integer;
  begin
  ConsoleForm.addLineToTranscript('> ' + commandPhrase, clRed);
  ConsoleForm.addLineToTranscript(ConsoleForm.speechSystem.stripMacros(newLastSaidTextWithMacros), clBlue);
  ConsoleForm.scrollTranscriptEndIntoView;
  ConsoleForm.speechSystem.sayTextWithMacros(newLastSaidTextWithMacros);

  {common with undo}
  ConsoleForm.speechSystem.lastSaidTextWithMacros := newLastSaidTextWithMacros;
  domain.world.firstCommandDoneForLastCommandPhrase := newFirstCommandDoneForLastCommandPhrase;
  if newFocus <> domain.world.emptyEntry then
    begin
  	oldFocus.setState(kAbsent);
  	domain.world.focus := newFocus;
  	newFocus.setState(kPresent);
    end;
  for i := 0 to changedVariables.count - 1 do
    TSChangedVariableWrapper(changedVariables[i]).doChange;
  self.updateForChanges;
  ConsoleForm.speechSystem.checkForSayOptionsMacro;
  inherited doCommand;
 end;

procedure TSDoCommandPhrase.undoCommand;
  var
  	i: integer;
    undoPhrase: string;
  begin
  ConsoleForm.addLineToTranscript('> undo', clRed);
//  undoPhrase := 'It is as if "' + commandPhrase + '" had never been said.';
  undoPhrase := '(You decide not to say "' + commandPhrase + '")';
  ConsoleForm.addLineToTranscript(undoPhrase, clBlue);
  ConsoleForm.scrollTranscriptEndIntoView;
  ConsoleForm.speechSystem.speakText(undoPhrase);
  
  ConsoleForm.speechSystem.lastSaidTextWithMacros := oldLastSaidTextWithMacros;
  domain.world.firstCommandDoneForLastCommandPhrase := oldFirstCommandDoneForLastCommandPhrase;
  for i := changedVariables.count - 1 downto 0 do
    TSChangedVariableWrapper(changedVariables[i]).undoChange;
  if newFocus <> domain.world.emptyEntry then
    begin
  	newFocus.setState(newFocusOldState);
  	domain.world.focus := oldFocus;
  	oldFocus.setState(oldFocusOldState);
    end;
  self.updateForChanges;
  inherited undoCommand;
  end;

procedure TSDoCommandPhrase.redoCommand;
  var
  	i: integer;
    redoPhrase: string;
  begin
  ConsoleForm.addLineToTranscript('> redo', clRed);
  redoPhrase := '(You decide to say "' + commandPhrase + '" anyway)';
  ConsoleForm.addLineToTranscript(redoPhrase, clBlue);
  ConsoleForm.scrollTranscriptEndIntoView;
  ConsoleForm.speechSystem.speakText(redoPhrase);

  {common with do}
  ConsoleForm.speechSystem.lastSaidTextWithMacros := newLastSaidTextWithMacros;
  domain.world.firstCommandDoneForLastCommandPhrase := newFirstCommandDoneForLastCommandPhrase;
  if newFocus <> domain.world.emptyEntry then
    begin
  	oldFocus.setState(kAbsent);
  	domain.world.focus := newFocus;
  	newFocus.setState(kPresent);
    end;
  for i := 0 to changedVariables.count - 1 do
    TSChangedVariableWrapper(changedVariables[i]).doChange;
  self.updateForChanges;
  inherited doCommand;
 end;

function TSDoCommandPhrase.description: string;
  begin
  result := 'command: ' + commandPhrase;
  end;

end.
